### ISSUE TITLE
Serializing EphemeralPublic

### ISSUE THREAD

**Issue by thanethomson** [dalek-cryptography/x25519-dalek#25]
We need to serialize `EphemeralPublic` (over [here](https://github.com/dalek-cryptography/x25519-dalek/blob/master/src/x25519.rs#L27)) to an array of bytes so we can share it between network participants.

I see that the underlying [`MontgomeryPoint`](https://github.com/dalek-cryptography/curve25519-dalek/blob/master/src/montgomery.rs#L67) has a [`to_bytes()`](https://github.com/dalek-cryptography/curve25519-dalek/blob/master/src/montgomery.rs#L100) function, but this is inaccessible due to the fact that `EphemeralPublic` doesn't expose any such function, and the underlying `MontgomeryPoint` in the `EphemeralPublic` tuple is private outside of its crate.

Is there some other way to access a `to_bytes()` function for the `EphemeralPublic` struct that has the same effect?

This was easier in older versions of `x25519-dalek`, but the older `subtle` dependency being yanked (as per #19) prevents us from continuing to use the older version of the package.

**Comment by Liamsi** [comment]
Adding 
```rust
impl EphemeralPublic {
    pub fn to_bytes(&self) -> [u8; 32] {
        self.0.to_bytes()
    }
}
```
to https://github.com/dalek-cryptography/x25519-dalek/blob/master/src/x25519.rs should help to make it easier to do this:

> Alice meows across the room, telling `alice_public` to Bob, and Bob loudly meows `bob_public` back to Alice. 

Or are we missing another way to do so?

**Comment by tarcieri** [comment]
@Liamsi I think you probably want:

```rust
fn as_bytes(&self) -> &[u8; 32]
```

...which borrows the value (and is consistent with `SharedSecret`)

**Comment by thanethomson** [comment]
Done now @tarcieri (in #26). One thing I'm not 100% sure of is why the existing `SharedSecret.as_bytes()` function returns a reference to a reference to a `[u8; 32]` over here: https://github.com/dalek-cryptography/x25519-dalek/blob/master/src/x25519.rs#L92

The `MontgomeryPoint.as_bytes()` function already returns a reference to the underlying bytes.

**Comment by zer0x64** [comment]
Hey! Another thing I have to ask. 
I know that the EphemeralSecret struct is not supposed to be serialized so people don't stupidly send them everywhere, but in my specific use case(related to FFI and bindings with other language), it would be very beneficial to have a way to get them as byte array, since with some language opaque structures are pretty hard to deal with.
Is there a way to do it that that I haven't seen yet? Is there any plan to somehow implement this(with and unsafe flag or something to discourage people to use it)?

**Comment by hdevalence** [comment]
> Done now @tarcieri (in #26). One thing I'm not 100% sure of is why the existing `SharedSecret.as_bytes()` function returns a reference to a reference to a `[u8; 32]` over here: https://github.com/dalek-cryptography/x25519-dalek/blob/master/src/x25519.rs#L92
> 
> The `MontgomeryPoint.as_bytes()` function already returns a reference to the underlying bytes.

Looks like there's no good reason for this, feel free to remove it.

**Comment by hdevalence** [comment]
> Hey! Another thing I have to ask.
> I know that the EphemeralSecret struct is not supposed to be serialized so people don't stupidly send them everywhere, but in my specific use case(related to FFI and bindings with other language), it would be very beneficial to have a way to get them as byte array, since with some language opaque structures are pretty hard to deal with.
> Is there a way to do it that that I haven't seen yet? Is there any plan to somehow implement this(with and unsafe flag or something to discourage people to use it)?

So, this isn't something I'm set on, just an initial point of view: my feeling is that for FFI, it's better to have simple, byte-oriented APIs, rather than ones that attempt to maintain more complex invariants.  Right now we have two levels of API, a bare byte-oriented one that matches the RFC, and a more complex one that uses Rust types to try to maintain invariants around ephemeral keys, etc. Those invariants can't really be maintained across FFI, so I wonder whether it would be better to use the "bare" API.

The only limitation of the lower-level API is that it doesn't currently support the fast edwards-arithmetic keygen, but we could consider adding a second function for that.

**Comment by hdevalence** [comment]
The core of this is fixed in #29 , released in 0.4.4
